include/yaclib/coroutine/await.hpp

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <yaclib/async/future.hpp>
+
+#if defined(__clang__)
+
+#include <experimental/coroutine>
+#define coro_ns std::experimental
+
+#else
+
+#include <coroutine>
+#define coro_ns std
+
+#endif
+
+#include <iostream>  // for test
+#include <tuple>
+
+namespace yaclib {
+
+template <typename... Ts>
+class Await final {
+ public:
+  Await() = delete;
+  Await(const Await&) = delete;
+  Await(Await&&) = delete;
+  Await& operator=(Await&&) = delete;
+  Await& operator=(const Await&) = delete;
+  ~Await() = default;
+  Await(Future<Ts>&... fs) : _fut_refs(fs...), _cntr() {
+    std::apply(
+        [&](Future<Ts>&... futures) {
+          ((futures.GetCore()->SetWait(_cntr)), ...);
+        },
+        _fut_refs);
+  }
+
+  bool await_ready() const noexcept {
+    return _cntr.resumed;
+  }
+
+  void await_resume() const noexcept {
+  }
+
+  bool await_suspend(coro_ns::coroutine_handle<> handle) {
+    _cntr.handle = handle;
+    _cntr.DecRef();
+    return !_cntr.resumed;
+  }
+
+ private:
+  struct CoroHandler : public util::IRef {
+    coro_ns::coroutine_handle<> handle;
+    bool resumed = false;
+  };
+  struct CoroResumeDeleter {
+    static void Delete(CoroHandler* handle) {
+      if (!handle) {
+        std::cerr << "CoroHandler is nullptr!\n";
+        std::terminate();
```

Comment:
Remove std:: terminate, use assert now
In near future @myannyax add our own assert